homeassistant/components/livisi/coordinator.py

```diff
@@ -0,0 +1,139 @@
+        )
+        self.config_entry: ConfigEntry = config_entry
+        self.hass: HomeAssistant = hass
+        self.aiolivisi: AioLivisi = aiolivisi
+        self.websocket: Websocket = Websocket(aiolivisi)
+        self.devices: list[dict[Any, Any]] = []
+        self.switch_devices: list = []
+        self.rooms: dict[str, Any] = {}
+        self.serial_number: str | None
+        self.controller_type: str | None
+        self.os_version: str | None
+        self.is_avatar: bool = False
+        self.port: str = ""
+
+    async def _async_update_data(self):
+        """Set device list."""
+        shc_devices: dict[Any, Any] = await self.async_get_devices()
+        for device in shc_devices:
+            if device in self.devices:
+                continue
+            if device["type"] == PSS_DEVICE_TYPE:
+                if device not in self.switch_devices:
+                    self.switch_devices.append(device)
+                    async_dispatcher_send(self.hass, LIVISI_DISCOVERY_NEW, device)
+            self.devices.append(device)
+
+    async def async_setup(self) -> None:
+        """Set up the Livisi Smart Home Controller."""
+        if bool(self.aiolivisi.livisi_connection_data) is False:
+            livisi_connection_data = {
+                "ip_address": self.config_entry.data["host"],
+                "password": self.config_entry.data["password"],
+            }
+
+            await self.aiolivisi.async_set_token(
+                livisi_connection_data=livisi_connection_data
+            )
+        controller_info = await self.aiolivisi.async_get_controller()
+
+        controller_data = controller_info
+        if controller_data.get("controllerType") == "Avatar":
+            self.port = AVATAR_PORT
+            self.is_avatar = True
+        else:
+            self.port = CLASSIC_PORT
+            self.is_avatar = False
+        self.serial_number = controller_data.get("serialNumber")
+        self.controller_type = controller_data.get("controllerType")
+        self.os_version = controller_data.get("osVersion")
+
+    async def async_get_devices(self) -> list:
+        """Set the discovered devices list."""
+        shc_devices = await self.aiolivisi.async_get_devices()
+        if bool(shc_devices) is False:
+            shc_devices = await self.aiolivisi.async_get_devices()
+        return shc_devices
+
+    async def async_get_pss_state(self, capability):
+        """Set the PSS state."""
+        response = await self.aiolivisi.async_get_pss_state(capability[1:])
+        if response is None:
+            return
+        on_state = response.get("onState")
+        return on_state.get("value")
+
+    async def async_set_all_rooms(self):
+        """Set the room list."""
+        response = await self.aiolivisi.async_get_all_rooms()
+
+        for available_room in response:
+            self.rooms[available_room.get("id")] = available_room.get("config").get(
```

Comment:
Can `"id"` be missing in `available_room`? In that case we would be overwriting the `None` key for all rooms with the same config.

Reply:
`available_room.get("config").get(...` will error if the first lookup doesn't find the `"config"` key.

Reply:
Every included device in the LIVISI configuration is assigned to a room. There is no device without a room in the LIVISI application, so the `id` cannot be missing in `available_room` .

Reply:
Then we shouldn't use `dict.get(key)` but `dict[key]`. Only use `dict.get` when we don't know if the key is in the dict.